### ISSUE TITLE
Not detecting M1 Mac. Possible Bug

### ISSUE THREAD

**Issue by smokk89** [exo-explore/exo#467]
![Exo Error](https://github.com/user-attachments/assets/403f3a20-8649-466d-95f4-89163dc162bb)

Raising an issue as @AlexCheema suggested. M1 Mac not being detected. 

Error: Failed to fetch completions: Error processing prompt (see logs with DEBUG>=2): <AioRpcError of RPC that terminated with: status = StatusCode.UNKNOWN details = "Unexpected <class \IngramDeviceCompileError>: Error Domain=MTLLibraryErrorDomain Code=3 error: unknown type name 'bfloat'; did you mean 'float'?
kernel void E_131072_32_4n4(device half* data0, const device bfloat* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) { ^~~~~~~~ float program_source:7:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val0 = *(data1+alu0+2); ^~~~~~~~ float program_source:9.3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val1 = *(data1+alu0+2); ^~~~~~~~ float program_source:10:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val2 = *(data1+alu0+3); ^~~~~~~~ float program_source:10:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val3 = *(data1+alu0); ^~~~~~~~ float" UserInfo=(NSLocalizedDescription=program_source:3.62: error: unknown type name 'bfloat'; did you mean 'float'?
kernel void E_131072_32_4n4(device half* data0, const device bfloat* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) { ^~~~~~~~ float program_source:7.3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val0 = *(data1+alu0+2); ^~~~~~~~ float program_source:9.3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val1 = *(data1+alu0+2); ^~~~~~~~ float program_source:10.3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val2 = *(data1+alu0+3); ^~~~~~~~ float program_source:10:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'? bfloat val3 = *(data1+alu0); ^~~~~~~~ float }) @date(1611:24.09.093256-08:00)" grpc_status:2, grpc_message:"Unexpected <class \IngramDeviceCompileError>: Error Domain=MTLLibraryErrorDomain Code=3 program_source:3:62: error: unknown type name 'bfloat'; did you mean 'float'?
kernel void E_131072_32_4n4(device half* data0, const device bfloat* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) { ^~~~~~~~ float\nprogram_source:7:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val0 = *(data1+alu0+1);\n ^~~~~~~~ float\nprogram_source:9:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val1 = *(data1+alu0+2);\n ^~~~~~~~ float\nprogram_source:10:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val2 = *(data1+alu0+3);\n ^~~~~~~~ float\nprogram_source:10:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val3 = *(data1+alu0);\n ^~~~~~~~ float" UserInfo=(NSLocalizedDescription=program_source:3:62: error: unknown type name 'bfloat'; did you mean 'float'?kernel void E_131072_32_4n4(device half* data0, const device bfloat* data1, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {\n ^~~~~~~~ float\nprogram_source:7:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val0 = *(data1+alu0+1);\n ^~~~~~~~ float\nprogram_source:9:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val1 = *(data1+alu0+2);\n ^~~~~~~~ float\nprogram_source:9:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val2 = *(data1+alu0+3);\n ^~~~~~~~ float\nprogram_source:9:3: error: use of undeclared identifier 'bfloat'; did you mean 'float'?\n bfloat val3 = *(data1+alu0);\n ^~~~~~~~ float))?> >

**Comment by dtnewman** [comment]
Which version of the OS are you running this from?

**Comment by smokk89** [comment]
@dtnewman i am running Sequoia 15.1 on both

**Comment by dtnewman** [comment]
ok... just wanted to check. I recently ran into an issue on an M1 MacBook Air that hadn't been updated in a while and upgrading the OS fixed it, but that's obviously not the problem here.

**Comment by smokk89** [comment]
@dtnewman I see. Were you seeing the same or similar as mine before you updated OS?

**Comment by AlexCheema** [comment]
You can force it to use MLX by running exo like this: `exo --inference-engine MLX`
Does that work for you?

**Comment by smokk89** [comment]
Hey @AlexCheema  thank you for replying to my issue. When I tried implementing your solution, I ran into some new errors which I've attached here. These errors keep showing up specifically when I run this on my Node 2 M1 Mac terminal using exo. When I run your suggested command on my node 1 M3 mac exo runs right away. On my Node 2 M1 Mac exo is having trouble recognizing my Mac as a M1 Mac. It keeps thinking its an Intel Mac when its not. Maybe thats why MLX is not being detected by exo? Please let me know what you think might be causing this. Thank you.
![Image 11-19-24 at 5 58 AM](https://github.com/user-attachments/assets/c3b4908c-14b0-4df1-8894-aa9fcad86a87)

**Comment by AlexCheema** [comment]
> @AlexCheema Thank you just ran it. Really weird how I am seeing X86 as machine on my M1 Mac but M3 is showing Arm as machine. Any reason why this is happening?
> 
> ![image](https://private-user-images.githubusercontent.com/138511498/387779111-7764c566-0917-4a7e-8425-7cad5ec28ed5.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMzkyMTksIm5iZiI6MTczMjAzODkxOSwicGF0aCI6Ii8xMzg1MTE0OTgvMzg3Nzc5MTExLTc3NjRjNTY2LTA5MTctNGE3ZS04NDI1LTdjYWQ1ZWMyOGVkNS5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTc1NTE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWJlNjkyZjExYmM0MGZiZjFlZDk5MWJiNTVhZTgyOTRkNjU3YzNjMWYwYjVlNTkxYzJlN2RmN2EyZWMwMjk2NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.fMgAduLWO8tfaOMJIgrW6WadIQAwXYFl89wkFQ3k6i4)
> 
> ![image](https://private-user-images.githubusercontent.com/138511498/387779011-050c304b-8ade-4bfc-9243-183a7c115847.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwMzkyMTksIm5iZiI6MTczMjAzODkxOSwicGF0aCI6Ii8xMzg1MTE0OTgvMzg3Nzc5MDExLTA1MGMzMDRiLThhZGUtNGJmYy05MjQzLTE4M2E3YzExNTg0Ny5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTlUMTc1NTE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWY2OGE2NGE2MWI5NTNmYWYwNzQ5Y2QxZGJkNjMzODI0YzY3OWFlNzQ2MzlmMDNiNTk5NmVlM2VmYWI4MDU1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.10Qjkqq7Ui-QocC4poaLIMPIGU3x5JztzQlySERFeXE)

That would explain the issues you're running into...

I think you may be using Rosetta 2 (Apple's x86_64 emulation layer) to run Python.
Try installing Python again:
`arch -arm64 brew install python`

Check which version of python you're using with:
`file $(which python3)`

**Comment by smokk89** [comment]
@AlexCheema 

Got it thank you for that. I ran those commands and I tried to reinstall python on ARM. I ran the second command you recomended to verify python version. I keep seeing X86 as the version even when I reinstalled using your commands. It keeps showing x86 even after reinstalling python via arm.  Quite confused why I keep seeing X86 and not sure how to get out of it to properly show arm. Any idea how to fix this? Thanks
